sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/FieldTypeDescriptors.java

```diff
@@ -0,0 +1,119 @@
+          .put(TypeName.INT64, TypeDescriptors.longs())
+          .put(TypeName.DECIMAL, TypeDescriptors.bigdecimals())
+          .put(TypeName.FLOAT, TypeDescriptors.floats())
+          .put(TypeName.DOUBLE, TypeDescriptors.doubles())
+          .put(TypeName.STRING, TypeDescriptors.strings())
+          .put(TypeName.DATETIME, TypeDescriptor.of(Instant.class))
+          .put(TypeName.BOOLEAN, TypeDescriptors.booleans())
+          .put(TypeName.BYTES, TypeDescriptor.of(byte[].class))
+          .build();
+  /** Get a {@link TypeDescriptor} from a {@link FieldType}. */
+  public static TypeDescriptor javaTypeForFieldType(FieldType fieldType) {
+    switch (fieldType.getTypeName()) {
+      case ARRAY:
+        return TypeDescriptors.lists(javaTypeForFieldType(fieldType.getCollectionElementType()));
+      case MAP:
+        return TypeDescriptors.maps(
+            javaTypeForFieldType(fieldType.getMapKeyType()),
+            javaTypeForFieldType(fieldType.getMapValueType()));
+      case ROW:
+        return TypeDescriptors.rows();
+      default:
+        return PRIMITIVE_MAPPING.get(fieldType.getTypeName());
+    }
+  }
+  /** Get a {@link FieldType} from a {@link TypeDescriptor}. */
+  public static FieldType fieldTypeForJavaType(TypeDescriptor typeDescriptor) {
+    if (typeDescriptor.isArray()
+        || typeDescriptor.isSubtypeOf(TypeDescriptor.of(Collection.class))) {
+      return getArrayFieldType(typeDescriptor);
+    } else if (typeDescriptor.isSubtypeOf(TypeDescriptor.of(Map.class))) {
+      return getMapFieldType(typeDescriptor);
+    } else if (typeDescriptor.isSubtypeOf(TypeDescriptor.of(Row.class))) {
+      throw new IllegalArgumentException(
+          "Cannot automatically determine a field type from a Row class"
+              + " as we cannot determine the schema. You should set a field type explicitly.");
+    } else {
+      TypeName typeName = PRIMITIVE_MAPPING.inverse().get(typeDescriptor);
+      if (typeName == null) {
+        throw new RuntimeException("Couldn't find field type for " + typeDescriptor);
+      }
+      FieldType fieldType = FieldType.of(typeName);
+      return fieldType;
+    }
+  }
+
+  private static FieldType getArrayFieldType(TypeDescriptor typeDescriptor) {
+    if (typeDescriptor.isArray()) {
+      if (typeDescriptor.getComponentType().getType().equals(byte.class)) {
+        return FieldType.BYTES;
+      } else {
+        return FieldType.array(fieldTypeForJavaType(typeDescriptor.getComponentType()));
+      }
+    }
+    if (typeDescriptor.isSubtypeOf(TypeDescriptor.of(Collection.class))) {
+      TypeDescriptor<Collection<?>> collection = typeDescriptor.getSupertype(Collection.class);
+      if (collection.getType() instanceof ParameterizedType) {
+        ParameterizedType ptype = (ParameterizedType) collection.getType();
+        java.lang.reflect.Type[] params = ptype.getActualTypeArguments();
+        checkArgument(params.length == 1);
```

Comment:
Does it have to be 1, does Beam explicitly support only simple collections? E.g. there are [less obvious generic things](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.KeySetView.html) that are collections